src/app/AttributeAccessInterface.h

```diff
@@ -41,23 +42,135 @@
+        }
+
+    private:
+        AttributeValueEncoder * mAttributeValueEncoder;
+    };
+
+    class AttributeEncodeState
+    {
+    public:
+        AttributeEncodeState() : mAllowPartialData(false), mCurrentEncodingListIndex(kInvalidListIndex) {}
+        bool AllowPartialData() const { return mAllowPartialData; }
+
+    private:
+        friend class AttributeValueEncoder;
+
+        bool mAllowPartialData              = false;
+        ListIndex mCurrentEncodingListIndex = kInvalidListIndex;
+    };
+
+    AttributeValueEncoder(AttributeReportIBs::Builder & aAttributeReportIBsBuilder, FabricIndex aAccessingFabricIndex,
+                          const ConcreteAttributePath & aPath, DataVersion aDataVersion,
+                          const AttributeEncodeState & aState = AttributeEncodeState()) :
+        mAttributeReportIBsBuilder(aAttributeReportIBsBuilder),
+        mAccessingFabricIndex(aAccessingFabricIndex), mPath(aPath), mDataVersion(aDataVersion), mEncodeState(aState)
     {}
 
     template <typename... Ts>
     CHIP_ERROR Encode(Ts... aArgs)
     {
         mTriedEncode = true;
-        if (mWriter == nullptr)
+
+        AttributeReportIB::Builder attributeReportIBBuilder = mAttributeReportIBsBuilder.CreateAttributeReport();
+        AttributeDataIB::Builder attributeDataIBBuilder     = attributeReportIBBuilder.CreateAttributeData();
+        AttributePathIB::Builder attributePathIBBuilder     = attributeDataIBBuilder.CreatePath();
+        attributePathIBBuilder.Endpoint(mPath.mEndpointId).Cluster(mPath.mClusterId).Attribute(mPath.mAttributeId);
+
+        // Encode the list index field if we are encoding a single element in an list.
+        if (mCurrentEncodingListIndex != kInvalidListIndex)
+        {
+            attributePathIBBuilder.ListIndex(mCurrentEncodingListIndex);
+        }
+
+        ReturnErrorOnFailure(attributePathIBBuilder.EndOfAttributePathIB().GetError());
+
+        ReturnErrorOnFailure(
+            TagBoundEncoder(attributeDataIBBuilder.GetWriter(), TLV::ContextTag(to_underlying(AttributeDataIB::Tag::kData)))
+                .Encode(std::forward<Ts>(aArgs)...));
+
+        attributeDataIBBuilder.DataVersion(mDataVersion);
+
+        ReturnErrorOnFailure(attributeDataIBBuilder.EndOfAttributeDataIB().GetError());
+        return attributeReportIBBuilder.EndOfAttributeReportIB().GetError();
+    }
+
+    template <typename... Ts>
+    CHIP_ERROR EncodeListItem(Ts... aArgs)
+    {
+        if (mCurrentEncodingListIndex == kInvalidListIndex)
         {
+            if (mEncodeState.mCurrentEncodingListIndex == kInvalidListIndex)
+            {
+                // Spec 10.5.4.3.1, 10.5.4.6 (Replace a list w/ Multiple IBs)
+                // Put a empty array before encoding the first array element for list chunking.
+                AttributeReportIB::Builder attributeReportIBBuilder = mAttributeReportIBsBuilder.CreateAttributeReport();
```

Comment:
Again, this is a _lot_ of code that is being done per-`Ts` but shouldn't be.

Reply:
Extracted most non-control common code to separate functions.